[sully] 리뷰용 PR입니다

devbattery opened the pull request.

Commit:
docs: 학습 계획 작성

Commit:
feat: 마크업 파일 업로드

블로그에 있던 양식을 참고하여 root 폴더에 바로 놨는데 이게 맞는지는 잘 모르겠다

Commit:
fix: 마크업 파일 제거

프로젝트 구조를 공부하고 보니 resources 폴더 내에 static 폴더로 마크업 파일이 존재한다는 것을 깨달았다.

Commit:
feat: 카카오 마크업 파일로 변경

원래 위치한 static 폴더는 미션에서 주어진 디자인이 아니었다.

Commit:
feat: template 파일인 mustache 적용 시도

mustache 플러그인 설치 후 templates 폴더 안에 파일이 위치시키게 했다. 나중을 위한 발판이다.

Commit:
fix: 기존 qna 폴더로 다시 수정

호눅스와의 클래스를 갖고 난 후 제일 쉬운 방법이 기존의 폴더 방식이라는 것을 알게 되어서 기본 코드에 있는 폴더로 다시 수정했다.

Commit:
feat: templates 폴더에 form.html, index.html 복사 후 추가

동적 회원가입 기능 구현을 위해 복사

Commit:
feat: 타임리프 설정 후 폴더 위치 변경

gradle과 각 html 파일에 타임리프 코드를 추가해주고 폴더 위치를 static에서 templates로 옮겼다.

Commit:
feat: 사용자가 전달한 값을 관리하는 User 클래스 생성

각 필드를 생성한 후 setter와 getter 생성

Commit:
fix: userId를 id로 변경

내가 한 것은 index였고, 여기서 해야 하는 것은 아이디 그 고유의 값이므로 String형으로 변경

Commit:
style: id의 변수명 변경

html 파일을 보다 보니 id가 아닌 userId로 작성되어 있는 것을 발견
변수명만 간단하게 변경해 줌

Commit:
feat: 회원가입 버튼을 누르면 form.html로 넘어가게끔 구현

이제 코드 중복을 제거하고 매핑하는 데 문제가 없도록 할 예정입니다.

Commit:
feat: 가장 하단에 위치한 중복 코드 분리

header.html이라는 네이밍

Commit:
feat: 가장 상단에 위치한 중복 코드 분리

header.html이라는 네이밍

Commit:
feat: footer와 header를 불러오도록 구현

th:replace를 이용하여 분리한 코드를 불러옴

Commit:
frag: navbar-collapse 코드 분리 후 적용

index.html에서 user/join으로 수정해줬던 그 코드이다. 모든 코드에서 중복이 되어서 다 수정했다.

Commit:
fix: "."을 제거하였고, footer로 되어 있던 것을 수정

</div>.로 되어 있어서 그 부분을 수정하였고
index.html에서 navbar-collapse.html를 불러와야 할 부분에 footer.html이 불러오도록 짜여져 있었음

Commit:
style: navbar-collapse에서 navbar-collapse1로 변수명 변경

navbar-collapse2를 위해 변경

Commit:
frag: navbar-collapse1 코드 분리 후 적용

navbar-collapse2와 똑같은 방식이다

Commit:
frag: subnav.html 코드 분리 후 적용

Commit:
frag: navbar-fixed-top.html 코드 분리 후 적용

Commit:
feat: 유저 리포지토리 인터페이스 생성

MemoryUserRepository에서 이 인터페이스를 구현

Commit:
feat: save 메서드의 반환값을 Optional<>로 변경

값이 안 들어올 수 있음

Commit:
feat: UserRepository를 구현하는 MemoryUserRepository 클래스 생성 후 save 메서드 구현

Map의 key는 userId, value는 User 객체를 받는 HashMap을 생성
단, 실무에서는 동시성 문제 때문에 ConcurrentHashMap을 사용하지만 db에서 없애줄 예정이므로 HashMap 사용
userMap에 넣어줄 때 이미 key값이 존재한다면 Optional의 empty를 반환하도록 함

Commit:
feat: find 역할의 기능 추가

UserId와 객체를 반환해주는 메서드 두개를 구현했다. 항상 userId는 null인 경우를 생각한다.

Commit:
fix: get 방식에서 post 방식으로 변경

Commit:
feat: UserRepository 객체를 관리하는 Service 생성

UserRepository 객체를 생성자로 받아와서 관리한다.
userId가 중복되는지 판별하고, 저장해주는 회원가입 기능을 갖고 있다.
user 전체를 찾는 기능도 갖고 있다.

Commit:
feat: 회원가입 할 때 입력받은 값들을 받아올 수 있도록 생성자 생성

model에 저장한 값을 User 클래스 안에 넣어줄 예정이다.

Commit:
feat: UserService 객체를 관리하도록 생성자 추가

Commit:
feat: 사용자 추가 기능 구현

전에 User와 UserForm을 수정한 이유가 바로 이것이다. 사용자를 받아서 users에 리다이렉트 되도록 한다.

Commit:
feat: 리다이렉트된 /users를 list.html에 리턴되도록 구현

이게 가능한 이유는 상대 경로로 넘겨줄 때 ViewResolver가 경로를 추가해주기 때문

Commit:
feat: model로 받아온 리스트를 각각 테이블에 전달

하지만 userId, name, email에 노란색 밑줄로 Cannot resolve가 뜨는데 정상적으로 넘어옴. 이 이유는 모르겠음

Commit:
fix: HTML 코드 중복 분리 수정

form에서만 header부분의 css 경로가 살짝 달랐고, navbar-fixed-top.html에서 빼먹었던 부분을 추가했다.

Commit:
feat: profile.html로 userId의 user 객체를 넘겨주는 메서드 추가

디버깅으로 어쩌구/users/입력한ID 입력하니 잘 넘어오는 것을 확인

Commit:
fix: 경로 수정

Commit:
fix: header 분리

header와 header2로 분리하여 각 파일에 맞는 frag로 정의했다

Commit:
fix: header2 경로 수정

예전과 똑같은 실수함. ../css로 변경

Commit:
frag: index와 index가 아닌 html 분리

index에서 접근할 때는 "/user/어쩌구"가 맞지만
거기로 이동한 후에는 "/어쩌구"가 맞다.
이렇게 하면 로그인으로 이동한 후에 계속 로그인을 클릭해도 에러가 뜨지 않는다.

Commit:
fix: redirect 이해 후 적용

redirect를 잘 이해하지 못 하고 있어서 Direct 메서드에서 그냥 /login에 가게 만드는 실수를 저질렀다.
직접 몇번 실험해 보면서 이럴 때 redirect를 쓰는 것임을 깨달았다.

AlbertImKr commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,84 @@
+    @GetMapping("/join")
+    public String joinFormDirect() {
+        return "redirect:user/join";
+    }
```
- /join과 /user/join
- /user/login 과 /logoin
- /list 관 /user/list 등
중복되여 개인적으로 필요없다고 생각합니다

devbattery replied:
감사합니다! 덕분에 중복을 많이 줄일 수 있었습니다

AlbertImKr commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,84 @@
+
+    @GetMapping("/user/join")
+    public String joinForm() {
+        return "user/form";
```
여기서 new JoinForm객체를 Model에 담아서 보내야 html form에 그것을 매칭하고 posting할때 controller에서 값을 가져올수 있다고 알고 있습니다. (확실하지는 않습니다.)
보낼 때
```
public String joinForm(Model model){
    model.add(new JoinForm);
    return "user/form";
}
```
```
<form th:object="${joinForm}">
   <input th:field="*{nickname}" id=nickname>
   ...
</form>

```

받을 때 
```
public String create(UserForm form) {
        User user = new User(form.getUserId(), form.getPassword(), form.getName(), 
```

devbattery replied:
Model과 Controller의 관계에 대해서 다시 한번 더 공부해보겠습니다!

AlbertImKr commented on src/main/resources/templates/user/list.html:
```diff
@@ -0,0 +1,41 @@
+                    <td th:text="${user.userId}"></td>
+                    <td th:text="${user.name}"></td>
+                    <td th:text="${user.email}"></td>
+                    <td><a href="#" class="btn btn-success" role="button">수정</a></td>
```
href="" 에  링크를 추가 할수 있습니다.
```
<a th:herf="@{/users/{id}(id = ${user.id})}"
```

devbattery replied:
name에 링크 추가하도록 하겠습니다 :)

AlbertImKr reviewed (commented):
아직 저도 잘 몰라서 많이 남길 것이 없네요.

Commit:
fix: 절대 경로로 수정

상대 경로로 했기 때문에 header2, for-index와 같은 파일들이 추가적으로 생성할 수밖에 없었다. 절대 경로로 수정한 이후로는 한 종류의 html 파일로만 적용이 가능하게 됐다

Commit:
fix: Direct 메서드 삭제

html에서 상대 경로로 전달해줬었기 때문에 상대경로/상대경로 이런 식으로 계속 경로 추가가 됐다. 하지만 절대 경로로 바꾸면 이 모든 것이 해결된다.

Commit:
feat: 회원가입 후 프로필로 넘어가는 기능 구현

회원가입 했을 때의 userId를 바탕으로 하여 경로에 추가해준다. 그러면 그 userID의 profile로 연결이 된다.

Commit:
feat: userId와 email 연결

profile.html에 model이 넘겨줬으니 그 정보를 바탕으로 출력

Commit:
fix: header2, for-index 없앤 결과 적용

header2를 없애고, for-index를 없앴으니 그것을 user 폴더 내의 파일과 index에도 적용시켜야 함

Commit:
fix: save 메서드의 반환 타입을 void로 변경

return 값이 존재할 이유가 아직까지는 없다.

Commit:
feat: 사용자 아이디를 클릭하면 그 프로필로 이동하게끔 구현

Posts의 각 게시물마다 작성자를 클릭하면 이처럼 되도록 구현 예정
